File: backend/app/core/permissions.py
from fastapi import HTTPException, status
from .enums import UserRole

ROUTE_PERMISSIONS = {
    "courses:read": [UserRole.ADMIN, UserRole.TEACHER, UserRole.STUDENT],
    "courses:write": [UserRole.ADMIN, UserRole.TEACHER],
    "enrollment:write": [UserRole.STUDENT],
    "assignments:write": [UserRole.ADMIN, UserRole.TEACHER],
    "attendance:write": [UserRole.ADMIN, UserRole.TEACHER],
    "grades:write": [UserRole.ADMIN, UserRole.TEACHER],
    "audit:read": [UserRole.ADMIN],
}

def check_permission(role: UserRole | str, permission: str) -> bool:
    allowed = ROUTE_PERMISSIONS.get(permission, [])
    try:
        normalized = role if isinstance(role, UserRole) else UserRole(role)
    except ValueError:
        return False
    return normalized in allowed

def require_permission(role: UserRole | str, permission: str) -> None:
    if not check_permission(role, permission):
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Permission denied")

File: frontend/src/constants/permissions.ts
import { UserRole } from './enums';
export const RoutePermissions = {
  courses: [UserRole.ADMIN, UserRole.TEACHER, UserRole.STUDENT],
  assignments: [UserRole.ADMIN, UserRole.TEACHER, UserRole.STUDENT],
  grades: [UserRole.ADMIN, UserRole.TEACHER, UserRole.STUDENT],
  attendance: [UserRole.ADMIN, UserRole.TEACHER, UserRole.STUDENT],
  audit: [UserRole.ADMIN],
} as const;
export const ButtonPermissions = {
  createCourse: [UserRole.ADMIN, UserRole.TEACHER],
  changeAttendance: [UserRole.ADMIN, UserRole.TEACHER],
  modifyGrade: [UserRole.ADMIN, UserRole.TEACHER],
  enrollCourse: [UserRole.STUDENT],
} as const;
